proxy/http2/Http2Stream.cc

```diff
@@ -822,56 +812,37 @@ Http2Stream::response_get_data_reader() const
 void
 Http2Stream::set_active_timeout(ink_hrtime timeout_in)
 {
-  active_timeout = timeout_in;
-  clear_active_timer();
-  if (active_timeout > 0) {
-    active_event = this_ethread()->schedule_in(this, active_timeout);
-  }
+  _timeout.set_active_timeout(timeout_in);
 }
 
 void
 Http2Stream::set_inactivity_timeout(ink_hrtime timeout_in)
 {
-  inactive_timeout = timeout_in;
-  if (inactive_timeout > 0) {
-    inactive_timeout_at = Thread::get_hrtime() + inactive_timeout;
-    if (!inactive_event) {
-      inactive_event = this_ethread()->schedule_every(this, HRTIME_SECONDS(1));
-    }
-  } else {
-    clear_inactive_timer();
-  }
+  _timeout.set_inactive_timeout(timeout_in);
 }
 
 void
-Http2Stream::cancel_inactivity_timeout()
+Http2Stream::cancel_active_timeout()
 {
-  set_inactivity_timeout(0);
+  _timeout.cancel_active_timeout();
 }
+
 void
-Http2Stream::clear_inactive_timer()
+Http2Stream::cancel_inactivity_timeout()
 {
-  inactive_timeout_at = 0;
-  if (inactive_event) {
-    inactive_event->cancel();
-    inactive_event = nullptr;
-  }
+  _timeout.cancel_inactive_timeout();
 }
 
-void
-Http2Stream::clear_active_timer()
+bool
+Http2Stream::is_active_timeout_expired(ink_hrtime now)
 {
-  if (active_event) {
-    active_event->cancel();
-    active_event = nullptr;
-  }
+  return _timeout.is_active_timeout_expired(now);
```

Comment:
What benefit do we get from delegating this?

If Http2Stream inherited NetTimeout, we would not need to use template and the check that requires C++20. And the interface of NetTimeout would be clearer. Currently it has functions that are indirectly called from ActivityCop and functions that are called from Http2Stream to update internal state. These two kinds of function would be separated as public functions for ActivityCop and protected functions for Http2Stream. Then ActivityCop would be able to call is_active_timeout_expired directly and we would not need this function here.

Reply:
To make ActivityCop without Template, you need a class that inherits VConnection & NetTimeout. When Http2Stream inherits that class, you'll see the diamond problem. Also, when you try to make ActivityCop based InactivityCop, your class hierarchy becomes more complicated.

You can solve the complicated class hierarchy. But using Template is easier than that.

Reply:
Which part in ActivityCop requires VConnection interface?

Reply:
If you are talking about VC_EVENTs, it's not checked on your code neither. And technically the class doesn't have to  be a VConnection. Continuation satisfies the requirement for handling events.

If that part is going to be a problem, you could add `on_timeout()` to the interface. It's more general and you would be able to use it for any timeout.

Reply:
Yeh, it's `Continuation`. 
Assuming you're saying NetTimeout doesn't need to inherit the Continuation by declaring `int on_timeout(int event, Event *e)` as a pure virtual function. It will work, but it allows derived class to implement it anyhow. Some will call `Continuation::handleEvent(int, void*)` correctly, but some will not.
I prefer to restrict how events are passed & handled.

Reply:
Calling `handleEvent` would not ensure that timeout is going to be handled correctly. You still need to implement the handler correctly. In this sense, having `handleEvent` is not enough, and I'd call it a halfway restriction. `on_timeout()` would be more robust and flexible interface because it does not depend on other mechanisms. Changes against how we handle timeout on VConnection does not affect ActivityCop's job, and like I said on the comment above, you can use AC for other timeouts.

Also, having `on_timeout` and having `handleEvent` at somewhere are not exclusive. You could implement `on_timeout` to `(Net)VConnection` if you wanted to make sure `handleEvent` is going to be called at least.